aesara/tensor/shape.py

```diff
@@ -396,136 +376,157 @@ class SpecifyShape(COp):
+            if (s is None or NoneConst.equals(s))
+            else at.as_tensor_variable(s, ndim=0)
+            for s in shape
+        )
 
-        if any(s.dtype not in aesara.tensor.type.integer_dtypes for s in shape):
+        if any(
+            s.dtype not in aesara.tensor.type.integer_dtypes
+            for s in shape
+            if hasattr(s, "dtype")
+        ):
             raise TypeError("Shape values must be integer types")
 
         if len(shape) != x.type.ndim:
             raise ValueError(
                 f"Input `x` is {x.type.ndim}-dimensional and will never match a shape of length {len(shape)}."
             )
 
-        if isinstance(x.type, TensorType) and all(isinstance(s, Number) for s in shape):
-            out_var = x.type.clone(shape=shape)()
-        else:
-            out_var = x.type()
+        type_shape = [None] * x.ndim
+        for i, (xts, s) in enumerate(zip(x.type.shape, shape)):
+            if xts is not None:
+                type_shape[i] = xts
+            else:
+                try:
+                    type_s = get_scalar_constant_value(s)
+                    if type_s is not None:
+                        type_shape[i] = int(type_s)
+                except NotScalarConstantError:
+                    pass
+
+        out_var = x.type.clone(shape=type_shape)()
 
-        in_shape = at.as_tensor_variable(shape, ndim=1)
-        return Apply(self, [x, in_shape], [out_var])
+        return Apply(self, [x, *shape], [out_var])
 
     def perform(self, node, inp, out_):
-        x, shape = inp
+        x, *shape = inp
         (out,) = out_
         ndim = len(shape)
         if x.ndim != ndim:
             raise AssertionError(
                 f"SpecifyShape: Got {x.ndim} dimensions (shape {x.shape}), expected {ndim} dimensions with shape {tuple(shape)}."
             )
-        if x.shape != tuple(shape):
+        if not all(xs == s for xs, s in zip(x.shape, shape) if s is not None):
             raise AssertionError(
-                f"SpecifyShape: Got shape {x.shape}, expected {tuple(shape)}."
+                f"SpecifyShape: Got shape {x.shape}, expected {tuple(int(s) if s is not None else None for s in shape)}."
             )
         out[0] = x
 
     def infer_shape(self, fgraph, node, shapes):
-        xshape, sshape = shapes
+        xshape, *_ = shapes
+        shape = node.inputs[1:]
         new_shape = []
         for dim in range(node.inputs[0].type.ndim):
+            s = shape[dim]
             try:
-                s = at.get_scalar_constant_value(node.inputs[1][dim])
-                s = at.as_tensor_variable(s)
-                new_shape.append(s)
+                s = at.get_scalar_constant_value(s)
+                # We assume that `None` shapes are always retrieved by
+                # `get_scalar_constant_value`, and only in that case do we default to
+                # the shape of the input variable
+                if s is None:
+                    s = xshape[dim]
```

Comment:
The failing rewrites were due to me being perhaps over-careful here and defaulting to using the shape of the input variable `xshape` whenever `at.get_scalar_constant_value` failed to return a Constant. I was doing this because I feared a hidden `None` specified shape could be lurking somewhere in the graph of the shape. 

However, since most Op's cannot build on top of `NoneConst` inputs I have now switched to assuming that any error also confirms that the specified shape is not `None` and can be used for shape inference in this `Op`. The rewrites now work, but let's see if anything else fails

Reply:
Seems to work